tests/Tokenizer/TokensTest.php

```diff
@@ -29,6 +29,587 @@
+                2,
+                '<?php            ',
+                '<?php            ;',
+            ],
+            '5/' => [
+                1,
+                '<?php ',
+                "<?php\n;",
+            ],
+            // --- before and after
+            [
+                3,
+                '<?php
+# A7
+# B7
+',
+                '<?php
+# A7
+;
+# B7
+',
+            ],
+            [
+                4,
+                '<?php
+                # A9
+                # B
+                ',
+                '<?php
+                # A9
+                ;        '.'
+                # B
+                ',
+            ],
+            // --- after token
+            '10/' => [
+                1,
+                '<?php ',
+                '<?php
+;     ',
+            ],
+            '11/' => [
+                2,
+                '<?php
+',
+                '<?php
+
+;     ',
+            ],
+            '12/' => [
+                1,
+                '<?php ',
+                '<?php
+;',
+            ],
+
+            '14/' => [
+                3,
+                '<?php
+# a',
+                '<?php
+# a
+;     ',
+            ],
+            '16/' => [
+                1,
+                '<?php
+ # 1
+',
+                '<?php
+; # 1
+',
+            ],
+            '17/a' => [
+                3,
+                '<?php
+# 17a
+ # 18a
+',
+                '<?php
+# 17a
+; # 18a
+',
+            ],
+            '17/b' => [
+                3,
+                '<?php
+# 17b
+  # 18b
+',
+                '<?php
+# 17b
+ ; # 18b
+',
+            ],
+            '17/c' => [
+                3,
+                '<?php
+# 17c
+ # 18c
+',
+                '<?php
+# 17c
+ ;# 18c
+',
+            ],
+            '17/d' => [
+                2,
+                '<?php
+/* 17d*/# 18d
+',
+                '<?php
+/* 17d*/;# 18d
+',
+            ],
+            // --- special cases
+            '18/' => [
+                0,
+                '',
+                '<?php ',
+            ],
+            '20/' => [
+                1,
+                '<?php ',
+                '<?php  ',
+            ],
+            '21/' => [
+                1,
+                '<?php ',
+                "<?php\n ",
+            ],
+            '22/' => [
+                2,
+                '<?php
+ '.'
+  '.'
+   '.'
+    '.'
+     '.'
+     '.'
+    '.'
+   '.'
+  '.'
+ ',
+                '<?php
+ '.'
+  '.'
+   '.'
+    '.'
+     '.'
+    # delete this line only 22
+     '.'
+    '.'
+   '.'
+  '.'
+ ',
+            ],
+        ];
+    }
+
+    /**
+     * @param int    $index    token index to clear
+     * @param string $expected PHP source expected after clearing
+     * @param string $input    PHP source input
+     *
+     * @dataProvider provideClearTokenMergeSurroundingWhitespaceRemoveLineIfLeftEmpty2Cases
+     */
+    public function testClearTokenMergeSurroundingWhitespaceRemoveLineIfLeftEmpty2($index, $expected, $input)
```

Comment:
Can we have one test test and compare tokens collection like here and not code like in the `testClearTokenMergeSurroundingWhitespaceRemoveLineIfLeftEmpty`?

Reply:
Here we cannot because the tests here show some edge cases causing hard to fix issues.
For example:
```
            'D3' => [
                0,
                " \necho 14;",
                "<?php\n \necho 14;",
            ],
```

By removing `<?php` all other tokens stay the same, however when re-tokenizing the new content " \necho 14;" these are different. This is not unique for the new method, it as a limitation of our Tokens model.